client/src/main/java/org/apache/celeborn/client/read/LocalPartitionReader.java

```diff
@@ -0,0 +1,206 @@
+    this.location = location;
+    this.endMapIndex = endMapIndex;
+    PbStreamHandler streamHandle;
+    long fetchTimeoutMs = conf.clientFetchTimeoutMs();
+    try {
+      TransportClient client =
+          clientFactory.createClient(location.getHost(), location.getFetchPort(), 0);
+      TransportMessage openStreamMsg =
+          new TransportMessage(
+              MessageType.OPEN_STREAM,
+              PbOpenStream.newBuilder()
+                  .setShuffleKey(shuffleKey)
+                  .setFileName(location.getFileName())
+                  .setStartIndex(startMapIndex)
+                  .setEndIndex(endMapIndex)
+                  .setReadLocalShuffle(true)
+                  .build()
+                  .toByteArray());
+      ByteBuffer response = client.sendRpcSync(openStreamMsg.toByteBuffer(), fetchTimeoutMs);
+      streamHandle = TransportMessage.fromByteBuffer(response).getParsedPayload();
+    } catch (IOException | InterruptedException e) {
+      throw new IOException(
+          "Read shuffle file from local file failed, partition location: "
+              + location
+              + " filePath: "
+              + location.getStorageInfo().getFilePath(),
+          e);
+    }
+
+    chunkOffsets = new ArrayList<>(streamHandle.getChunkOffsetsList());
+    numChunks = streamHandle.getNumChunks();
+    shuffleChannel = FileChannelUtils.openReadableFileChannel(streamHandle.getFullPath());
+
+    logger.debug(
+        "Local partition reader {} offsets:{}",
+        location.getStorageInfo().getFilePath(),
+        StringUtils.join(chunkOffsets, ","));
+
+    if (hasPendingFetchTask.compareAndSet(false, true)) {
+      readLocalShufflePool.submit(() -> fetchDataTask());
+    }
+    ShuffleClient.incrementLocalReadCounter();
+  }
+
+  private void fetchDataTask() {
+    try {
+      while (!closed && currentChunkIndex < numChunks && results.size() < fetchMaxReqsInFlight) {
+        long offset = chunkOffsets.get(currentChunkIndex);
+        long length = chunkOffsets.get(currentChunkIndex + 1) - offset;
+        logger.debug("Read {} offset {} length {}", currentChunkIndex, offset, length);
+        // A chunk must be smaller than INT.MAX_VALUE
+        ByteBuffer buffer = ByteBuffer.allocate((int) length);
+        if (endMapIndex != Integer.MAX_VALUE) {
+          // skew partition
+          shuffleChannel.position(offset);
+        }
+        while (buffer.hasRemaining()) {
+          if (-1 == shuffleChannel.read(buffer)) {
+            exception.set(
+                new CelebornIOException(
+                    "Read local file " + location.getStorageInfo().getFilePath() + " failed"));
+          }
+        }
+        buffer.flip();
+        results.put(Unpooled.wrappedBuffer(buffer));
+        int oldChunkIndex = currentChunkIndex++;
+        logger.debug("Add index {} to results", oldChunkIndex);
+      }
+    } catch (Exception e) {
+      logger.warn("Read thread is cancelled.", e);
```

Comment:
I think we should not ignore the Exception here, because `shuffleChannel.position` can throw IOException. Better to `setException`.